### ISSUE TITLE
The cca loss of two irrelevant dataset is -20.

### ISSUE THREAD

**Issue by Umaruchain** [jameschapman19/cca_zoo#164]
Dear author,
Thanks for your great work. I am confused that the cca loss of two dataset is -20, which are produced by torch.randn((N,M)) separately.
I think the value should be 0 ?

**Comment by jameschapman19** [comment]
I understand you have trained a DCCA to convergence on some random data? The model will find correlations. A bit like how you could train a regression model with random data and a deep model.

**Comment by jameschapman19** [comment]
Your test loss should be ~0 just like a regression model should be approximately random out of sample with random input

**Comment by Umaruchain** [comment]
Here is the code. I do not train or use a model and I just calculate the loss function. The loss should be the -correlation but I am confused that why this is not around 0 but -18~-20? 
from cca_zoo.models import CCA
import numpy as np
import pdb
import torch

def _mat_pow(mat, pow_, epsilon):
    # Computing matrix to the power of pow (pow can be negative as well)
    [D, V] = torch.linalg.eigh(mat)
    mat_pow = V @ torch.diag((D + epsilon).pow(pow_)) @ V.T
    mat_pow[mat_pow != mat_pow] = epsilon  # For stability
    return mat_pow


def _demean(views):
    return tuple([view - view.mean(dim=0) for view in views])

def cca_loss(views,dim):
    n = views[0].shape[0]
        # Subtract the mean from each output
    views = _demean(views)

        # Concatenate all views and from this get the cross-covariance matrix
    all_views = torch.cat(views, dim=1)
    C = all_views.T @ all_views / (n - 1)
        #pdb.set_trace()

    # Get the block covariance matrix placing Xi^TX_i on the diagonal
    D = torch.block_diag(
            *[
                (1 - 0.0) * m.T @ m / (n - 1)
                + 0.0 * torch.eye(m.shape[1], device=m.device)
                for i, m in enumerate(views)
            ]
    )
    # pdb.set_trace()
    C = C - torch.block_diag(*[view.T @ view / (n - 1) for view in views]) + D

    R = _mat_pow(D, -0.5, 1e-3)

    # In MCCA our eigenvalue problem Cv = lambda Dv
    C_whitened = R @ C @ R.T

    eigvals = torch.linalg.eigvalsh(C_whitened)

    # Sort eigenvalues so lviewest first
    idx = torch.argsort(eigvals, descending=True)

    eigvals = eigvals[idx[: dim]]

    # leaky relu encourages the gradient to be driven by positively correlated dimensions while also encouraging
    # dimensions associated with spurious negative correlations to become more positive
    eigvals = torch.nn.LeakyReLU()(eigvals[torch.gt(eigvals, 0)] - 1)

    corr = eigvals.sum()

    return -corr

case = [torch.randn((2000,100)),torch.randn(2000,100)]

print(cca_loss(case,dim=100))

**Comment by jameschapman19** [comment]
yep this is expected. The analogy would again be a regression. If you generated X and y randomly it would be extremely unlikely that the r^2 would be zero.

**Comment by jameschapman19** [comment]
Also as n->infinity the r^2 would drop and likewise for CCA

**Comment by jameschapman19** [comment]
in fact for CCA when n=the number of features you will find perfect correlation. This is the reason why you will see many examples of people getting nans when training DCCA with batch size~=number of latent dimensions.

**Comment by jameschapman19** [comment]
you can actually experiment with this and prove it to yourself using your code :)

**Comment by Umaruchain** [comment]
wow!!! Yes, the loss is almost zero when I increase n to 20000. When I decrease n to 100, the loss is -98. So the batch size is critical when I train the DCCA. By the way, I have tried your model, DCCA_EigenGame, and it does not work very well when I set the batch size as the length of the whole dataset. Should I shuffle the training dataset or anything?

**Comment by jameschapman19** [comment]
I'd guess this is just learning rate

**Comment by jameschapman19** [comment]
I'd guess too big

**Comment by Umaruchain** [comment]
Yes... because the learning rate of the original dcca is 0.05 considering the batch size is 2000 and it looks fine. And the Egame does not work well when the rate is 0.05 or 0.005. I will try to set it smaller, many thanks!!!

**Comment by jameschapman19** [comment]
There isn't a theoretical reason why it should perform *better* than original DCCA in the full batch case but it should be much better as the batch size gets smaller.

**Comment by Umaruchain** [comment]
Yes!! That is what your job focuses on. Thanks again for your patience and help.
